File: 백준/Gold/2580. 스도쿠/스도쿠.py
import sys
array = [[] for _ in range(9)]
for i in range(9):
    array[i] = list(map(int,input().split()))

visited_row = [[False] * 10 for _ in range(9)]
visited_column = [[False] * 10 for _ in range(9)]
visited_squre = [[False] * 10 for _ in range(9)]

# 빈 칸 좌표 넣기
blank_list = []
for i in range(9):
    for j in range(9):
        num = array[i][j]
        if num == 0:
            blank_list.append((i,j))
        else:
            visited_row[i][num] = True
            visited_column[j][num] = True
            box_idx = (i // 3) * 3 + (j // 3)
            visited_squre[box_idx][num] = True
            
                
def recursive(depth):
    # basis part -> 모든 칸에 값을 넣었을 때 종료
    if depth == len(blank_list):
        for row in array:
            print(*row)
        sys.exit()
        
    # inductive part
    y, x = blank_list[depth]
    box_num = (y // 3) * 3 + (x // 3)

    # 넣을 수 있는 숫자 찾기
    for i in range(1, 10):
        if not visited_row[y][i] and not visited_column[x][i] and not visited_squre[box_num][i]:
            array[y][x] = i
            visited_row[y][i] = True
            visited_column[x][i] = True
            visited_squre[box_num][i] = True

            recursive(depth + 1)

            array[y][x] = 0
            visited_row[y][i] = False
            visited_column[x][i] = False
            visited_squre[box_num][i] = False
            
recursive(0)

File: 백준/Gold/2580. 스도쿠/스도쿠.cc
#include <iostream>
#include <vector>
using namespace std;

void recursive(int depth);

vector<pair<int, int>> blankList;

int array[9][9];
bool visitedRow[10][9];
bool visitedColumn[10][9];
bool visitedSquare[10][9];

int main() {
    for (int i = 0; i < 9; i++) {
        for (int j = 0; j < 9; j++) {
            cin >> array[i][j];
        }
    }

    // visited 배열 초기화
    for (int i = 0; i < 9; i++) {
        for (int j = 0; j < 10; j++) {
            visitedRow[i][j] = false;
            visitedColumn[i][j] = false;
            visitedSquare[i][j] = false;
        }
    }

    // 값이 0인 좌표 넣기
    for (int i = 0; i < 9; i++) {
        for (int j = 0; j < 9; j++) {
            int num = array[i][j];
            if (num == 0) {
                blankList.push_back({i, j});
            } else {
                visitedRow[i][num] = true;
                visitedColumn[j][num] = true;
                int squareBox = (i / 3) * 3 + (j / 3);
                visitedSquare[squareBox][num] = true;
            }
        }
    }

    recursive(0);
    return 0;
}

void recursive(int depth) {
    // basis part
    if (depth == blankList.size()) {
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                cout << array[i][j] << " ";
            }
            cout << "\n";
        }
        exit(0);
    }

    // inductive part -> 숫자 탐색
    int y = blankList[depth].first;
    int x = blankList[depth].second;
    int squareBox = (y / 3) * 3 + (x / 3);

    for (int i = 1; i < 10; i++) {
        if (!visitedRow[y][i] && !visitedColumn[x][i] && !visitedSquare[squareBox][i]) {
            array[y][x] = i;
            visitedRow[y][i] = true;
            visitedColumn[x][i] = true;
            visitedSquare[squareBox][i] = true;

            recursive(depth + 1);

            // 돌리기
            array[y][x] = 0;
            visitedRow[y][i] = false;
            visitedColumn[x][i] = false;
            visitedSquare[squareBox][i] = false;
        }
    }
}
